### DISCUSSION TITLE
[Suggestion]when the query rows is bigger than 1000...

### DISCUSSION THREAD

**Discussion by HaveF** [rclement/datasette-dashboards#142] [Ideas]
[The default query max rows is 1000](https://docs.datasette.io/en/latest/cli-reference.html#datasette-serve-help-settings)
```
  max_returned_rows            Maximum rows that can be returned from a table or
                               custom query (default=1000)
```

However, the number of points on the chart can exceeds 1000 easily. While we have the option to increase the number of rows, it would be beneficial to display a tooltip for points that exceed this limit. This would help viewers understand that they are only seeing a subset of the query results... 

![image](https://github.com/rclement/datasette-dashboards/assets/54462/317ceeec-e8f7-48c9-bce7-3a520e46cd45)

If the number of query rows exceeds the maximum number of rows to be returned, a helpful hint will appear. Simply move the mouse upwards to reveal the hints. What do you think?
![image](https://github.com/rclement/datasette-dashboards/assets/54462/b30db206-2c6a-404b-8f94-a1d752aec8c2)

**Comment by rclement** [comment]
Oh good catch! Even though most charts will make use of aggregated metrics (way less than a thousand data points) this is still an issue to deal with.

I cannot find a way for queries (CSV, JSON) to return the truncation information so that charts could display a warning. Do you have some pointers?

**Reply by rclement** [reply]
Ok so using the JSON API with `?_shape=objects` instead of `?_shape=array`, we can extract the truncation information:

```
https://datasette-dashboards-demo.vercel.app/jobs.json?sql=SELECT date(date) as Date, title as Title, company as Company, source as Source, city as City, longitude, latitude FROM offers_view WHERE longitude IS NOT NULL AND latitude IS NOT NULL AND date >= date(:date_start) &date_start=2021-01-01&_shape=objects
```

```json
{
    "ok": true,
    "database": "jobs",
    "query_name": null,
    "rows": [
        {
            "Date": "2021-04-21",
            "Title": "Devops Python H/F",
            "Company": "Sully Group",
            "Source": "RegionsJob",
            "City": "Aix-en-Provence",
            "longitude": 5.4451153,
            "latitude": 43.5227288
        },
        ...
    ],
    "truncated": true,
    "columns": [
        "Date",
        "Title",
        "Company",
        "Source",
        "City",
        "longitude",
        "latitude"
    ],
    "query": {
        "sql": "SELECT\n  date(date) as Date,\n  title as Title,\n  company as Company,\n  source as Source,\n  city as City,\n  longitude,\n  latitude\nFROM offers_view WHERE longitude IS NOT NULL AND latitude IS NOT NULL\n   AND date >= date(:date_start) \n  \n  \n  ",
        "params": {
            "date_start": "2021-01-01"
        }
    },
    "error": null,
    "private": false,
    "allow_execute_sql": true,
    "query_ms": 6.713455999943108
}
```

Charts are currently using either the CSV API or the JSON API (`_shape=array` and `_shape=arrays`) for convenience reasons. All charts would have to be converted to using exclusively the JSON API with `_shape=objects` to retrieve the truncation boolean and still obtain some manageable data objects from JavaScript.
 
However, let's keep in mind the JSON API might change with Datasette 1.0: https://github.com/simonw/datasette/milestone/7

**Reply by rclement** [reply]
As a comparison reference, here is how Metabase handles that:
- A row limit of 2000 by default for unaggregated data
- A row limit of 10000 for aggregated data
- All charts display a small warning tooltip when the limit is exceeded
- The row limit can be customized from the UI (a recent addition, it used to be hardcoded)

Here are some screenshots showing the unified warning tooltip for all chart types:

<img width="550" alt="Screenshot 2023-08-08 at 11 02 32" src="https://github.com/rclement/datasette-dashboards/assets/1238873/266b2c12-6f60-43aa-932e-4d6e693b3fde">

<img width="550" alt="Screenshot 2023-08-08 at 11 03 42" src="https://github.com/rclement/datasette-dashboards/assets/1238873/0b10534f-7e37-47ed-9e33-2ad287eef7aa">

**Reply by rclement** [reply]
I'll open an issue to address this

**Reply by rclement** [reply]
Issue #144

**Reply by HaveF** [reply]
I don't know how datasettle check this. What I notice is here:

https://datasette-dashboards-demo.vercel.app/jobs?sql=SELECT+date%28date%29+as+day%2C+count%28%2A%29+as+count+FROM+offers_view+WHERE+TRUE++AND+date+%3E%3D+date%28%3Adate_start%29+++++GROUP+BY+day+ORDER+BY+day&date_start=2021-01-01

![image](https://github.com/rclement/datasette-dashboards/assets/54462/4d10056d-8f52-4a54-b727-cb8650af93b6)
If you change the SQL, you will get:

![image](https://github.com/rclement/datasette-dashboards/assets/54462/4f9c9f69-14d1-4133-ade4-eee20ca3634e)

The "more than" part [is here](https://github.com/simonw/datasette/blob/86e4b3d33685a31668bd4ecc56e403e419940fe6/datasette/templates/query.html#L36)

**Comment by rclement** [comment]
Resolved with PR #156
